routing/index_graph.cpp

```diff
@@ -13,27 +16,520 @@ IndexGraph::IndexGraph(unique_ptr<GeometryLoader> loader, shared_ptr<EdgeEstimat
+  GetOutgoingGeomEdges(outgoingEdges, centerId, outgoingGeomEdges);
+
+  vector<RoadPoint> ingoingPath;
+  GetFeatureConnectionPath(restrictionInfo.m_from, centerId, restrictionInfo.m_fromFeatureId,
+                           ingoingPath);
+  if (ingoingPath.empty())
+    return;
+
+  JointEdgeGeom ingoingEdge(restrictionInfo.m_from, ingoingPath);
+  CHECK(!ingoingEdge.GetPath().empty(), ());
+
+  Joint::Id newJoint = Joint::kInvalidId;
+  uint32_t ingoingFeatureId = 0;
+  uint32_t outgoingFeatureId = 0;
+  for (auto it = outgoingGeomEdges.cbegin(); it != outgoingGeomEdges.cend(); ++it)
+  {
+    CHECK(!ingoingEdge.GetPath().empty(), ());
+    // Adding features 4->N and N->1 in example above.
+    if (it == outgoingGeomEdges.cbegin())
+    {      
+      if (restrictionInfo.m_from == centerId || centerId == it->GetTarget())
+      {
+        // @TODO(bykoianko) In rare cases it's posible that outgoing edges staring from |centerId|
+        // contain as targets |centerId|. The same thing with ingoing edges.
+        // The most likely it's a consequence of adding
+        // restrictions with type no for some bidirectional roads. It's necessary to
+        // investigate this case, to undestand the reasons of appearing such edges clearly,
+        // prevent appearing of such edges and write unit tests on it.
+        return;
+      }
+
+      // Ingoing edge.
+      ingoingFeatureId = AddFakeLooseEndFeature(restrictionInfo.m_from, ingoingEdge.GetPath(),
+                                                GetRoad(restrictionInfo.m_fromFeatureId).GetSpeed());
+      newJoint =
+          InsertJoint({ingoingFeatureId, static_cast<uint32_t>(ingoingEdge.GetPath().size() - 1)});
+
+      // Edge mapping.
+      InsertToEdgeMapping(from, DirectedEdge(restrictionInfo.m_from, newJoint, ingoingFeatureId));
+    }
+
+    outgoingFeatureId = AddFakeFeature(newJoint, it->GetTarget(), it->GetPath(),
+                                       GetSpeed(it->GetPath().front()));
+    // Edge mapping.
+    DirectedEdge const toItEdge(centerId, it->GetTarget(), it->GetPath().front().GetFeatureId());
+    InsertToEdgeMapping(toItEdge, DirectedEdge(newJoint, it->GetTarget(), outgoingFeatureId));
+  }
+
+  DisableEdge(from);
+}
+
+void IndexGraph::ApplyRestrictionOnlyRealFeatures(RestrictionPoint const & restrictionPoint)
+{
+  ApplyRestrictionRealFeatures(restrictionPoint, [&] (RestrictionInfo const & restrictionInfo) {
+    ApplyRestrictionOnly(restrictionInfo);
+  });
+}
+
+void IndexGraph::ApplyRestrictionOnly(RestrictionInfo const & restrictionInfo)
+{
+  Joint::Id centerId = restrictionInfo.m_center;
+
+  if (restrictionInfo.m_to == centerId || restrictionInfo.m_from == centerId)
+    return;
+
+  vector<JointEdge> ingoingEdges;
+  vector<JointEdge> outgoingEdges;
+  if (!GetIngoingAndOutgoingEdges(centerId, false, ingoingEdges, outgoingEdges))
+    return;
+
+  // One outgoing edge case.
+  if (outgoingEdges.size() == 1)
```

Comment:
I guess yes:
- we don't add some redundant edges and fake features;
- sometimes it's easier to what's going on;
- this cases require very small amount of code;